axum-macros/src/from_request.rs

```diff
@@ -0,0 +1,381 @@
+use proc_macro2::TokenStream;
+use quote::{quote, quote_spanned};
+use syn::{
+    parse::{Parse, ParseStream},
+    punctuated::Punctuated,
+    spanned::Spanned,
+    Token,
+};
+
+const GENERICS_ERROR: &str = "`#[derive(FromRequest)] doesn't support generics";
+
+pub(crate) fn expand(item: syn::ItemStruct) -> syn::Result<TokenStream> {
+    let syn::ItemStruct {
+        attrs,
+        ident,
+        generics,
+        fields,
+        semi_token: _,
+        vis,
+        struct_token: _,
+    } = item;
+
+    if !generics.params.is_empty() {
+        return Err(syn::Error::new_spanned(generics, GENERICS_ERROR));
+    }
+
+    if let Some(where_clause) = generics.where_clause {
+        return Err(syn::Error::new_spanned(where_clause, GENERICS_ERROR));
+    }
+
+    let FromRequestAttrs { via } = parse_attrs(&attrs)?;
+
+    if let Some((_, path)) = via {
+        impl_by_extracting_all_at_once(ident, fields, path)
+    } else {
+        impl_by_extracting_each_field(ident, fields, vis)
+    }
+}
+
+fn impl_by_extracting_each_field(
+    ident: syn::Ident,
+    fields: syn::Fields,
+    vis: syn::Visibility,
+) -> syn::Result<TokenStream> {
+    let extract_fields = extract_fields(&fields)?;
+
+    let (rejection_ident, rejection) = if let syn::Fields::Unit = &fields {
+        (syn::parse_quote!(::std::convert::Infallible), quote! {})
+    } else {
+        let rejection_ident = rejection_ident(&ident);
+        let rejection = extract_each_field_rejection(&ident, &fields, &vis)?;
+        (rejection_ident, rejection)
+    };
+
+    Ok(quote! {
+        #[::axum::async_trait]
+        impl<B> ::axum::extract::FromRequest<B> for #ident
```

Comment:
These impls should be marked [automatically_generated](https://doc.rust-lang.org/reference/attributes/derive.html#the-automatically_derived-attribute).

```suggestion
        #[::axum::async_trait]
        #[automatically_generated]
        impl<B> ::axum::extract::FromRequest<B> for #ident
```

Reply:
Uhh not heard about `#[automatically_generated]` before! Where is that documented?

Reply:
At the place I linked to 😅

It can also improve compile times, this is the first reference to that I could find: https://github.com/rust-lang/rust/commit/24d410abd6924c175ee776ed84be851b615a555b

Reply:
oh lol sorry 😜 cool!

Reply:
Seems to be `#[automatically_derived]`

Reply:
Oh, right ^^